### ISSUE TITLE
Move template code to header

### ISSUE THREAD

**Issue by innerlee** [pytorch/pytorch#98907]
### 🐛 Describe the bug

The template code https://github.com/pytorch/pytorch/blob/39fd7f945f292bea7af411946f75417966470359/torch/csrc/api/src/nn/modules/batchnorm.cpp#L17-L34 appears in the `.cpp` file.
It will make extending from `BatchNormImplBase` using libtorch troublesome.
Unless copying the above quoted code into users own source code, the linker will not able to find implementations of `pretty_print`

### Versions

latest master
